File: extractors/go/imports.go
// imports.go — IMPORTS edges + [mod] third-party nodes + the local-package
// membership index. No PACKAGE nodes (removed 2026-08-17 — engine-wide
// connected-only pruning makes package stubs obsolete; import edges now
// target the imported package's FILES, matching JS/Java/Rust).
//
// Resolution (documented Go semantics): an import path is internal when it
// starts with the module path (honoring `replace` directives, including
// filesystem replaces) — otherwise external → [mod]/... node (gated). std
// packages (no dot in first path element) → no node, no edge. Go imports
// PACKAGES, so the IMPORTS edge fans out from the importing FILE to every
// non-test FILE of the target package (one edge per file; package identity
// rides in each file's metadata.package).

package main

import (
	"path/filepath"
	"strings"
)

// detectImports — returns IMPORTS edges + seeds the shared ThirdPartyRegistry
// (imports/calls/inheritance share one gated registry per extraction) and the
// PkgNodesByPath membership index (local-vs-external checks in
// calls/routes/orm/inheritance — key presence only, values unused).
func detectImports(pr *parsedRepo, l *LookupMaps, opts *Options, fp *Fingerprint) []map[string]any {
	edges := []map[string]any{}

	l.tp = newThirdPartyRegistry(opts.IncludeThirdPartyLibs, fp, l)

	// Index: import path → non-test files of that package (FILE→FILE targets),
	// plus import path → dir for filesystem-replace resolution.
	filesByImportPath := map[string][]string{}
	dirByImportPath := map[string]string{}
	for _, pkg := range pr.packages {
		l.PkgNodesByPath[pkg.ImportPath] = nil // membership marker only
		relFiles := []string{}
		for _, pf := range pkg.Files {
			if pf.IsTest {
				continue // test files are never import targets (leaf rule)
			}
			relFiles = append(relFiles, pf.RelPath)
		}
		filesByImportPath[pkg.ImportPath] = relFiles
		dirByImportPath[pkg.ImportPath] = pkg.Dir
	}
	// replace-directive aliases: import path → local dir (filesystem replaces)
	fsReplaceDir := map[string]string{}
	for old, newp := range pr.mod.Replaces {
		if strings.HasPrefix(newp, "./") || strings.HasPrefix(newp, "../") || strings.HasPrefix(newp, "/") {
			fsReplaceDir[old] = strings.TrimPrefix(filepath.ToSlash(newp), "./")
		}
	}

	for _, pf := range pr.files {
		fileID := "file::" + pf.RelPath
		for _, imp := range pf.Imports {
			p := imp.Path
			if isStdImport(p) {
				continue // std: no node, no edge
			}
			// internal? (a) module-prefixed → package files, (b) replace →
			// filesystem dir → that package's files
			targets, ok := filesByImportPath[p]
			if !ok {
				if dir, ok := fsReplaceDir[p]; ok {
					for ip, d := range dirByImportPath {
						if strings.TrimPrefix(filepath.ToSlash(d), "./") == dir {
							targets = filesByImportPath[ip]
							ok = true
							break
						}
					}
				}
			}
			if ok {
				for _, rel := range targets {
					edges = append(edges, edgeWithMeta(fileID, "file::"+rel, EdgeImports,
						map[string]any{"importPath": p}))
				}
				continue
			}
			// external → [mod]/... (gated)
			if n := l.tp.packageNode(p); n != nil {
				edges = append(edges, edgeWithMeta(fileID, n["id"].(string), EdgeImports,
					map[string]any{"importPath": p, "isThirdParty": true}))
			}
		}
	}
	return edges
}

File: extractors/rust/src/imports.rs
// imports.rs — IMPORTS edges from `use` declarations + [crate] third-party
// nodes (mirrors go/imports.go).
//
// Resolution (documented Rust semantics — the module tree, not a sys.path
// simulation):
//   crate::a::b  → absolute;  super::/self::  → relative;  bare → relative
//   then crate root. Longest module prefix wins; the remainder is an item.
// Targets are FILES (Python model — no MODULE nodes in V1):
//   `use crate::models::user::User;` → IMPORTS edge to src/models/user.rs
// External first segments (a Cargo.toml dep, not std, not local) →
//   gated [crate]/name[::member] node + IMPORTS edge to it. Nodes are
//   registered into the shared ThirdPartyRegistry (the pipeline collects
//   tp.all_nodes() last — go parity); this module only emits edges.

use crate::contract::CodeEdge;
use crate::extractor::{Options, ParsedRepo};
use crate::fingerprint::ParsedManifest;
use crate::module_map::ModuleMap;
use crate::thirdparty::ThirdPartyRegistry;

pub struct ImportsOutput {
    pub edges: Vec<CodeEdge>,
}

/// Emit IMPORTS edges for every use declaration.
pub fn detect_imports(
    repo: &ParsedRepo,
    mf: &ParsedManifest,
    module_map: &ModuleMap,
    tp: &mut ThirdPartyRegistry,
    _opts: &Options,
) -> ImportsOutput {
    let mut edges = vec![];
    let mut seen: std::collections::HashSet<(String, String)> = std::collections::HashSet::new();

    for pf in &repo.files {
        if pf.is_test_file {
            // test files ARE leaf nodes, but they still import — emit edges
            // from the TEST node (same file id — parentFile covers it)
        }
        let from = format!("file::{}", pf.rel_path);
        let base_module = crate::walker::module_path_for_file(&pf.rel_path);
        for u in &pf.uses {
            // resolve the target path against the module map (use-semantics:
            // crate-root absolute for bare paths — edition 2018)
            if let Some((file, rest)) = module_map.resolve_use(&u.target, &base_module) {
                // skip self-imports (`use super::*` inside a same-file mod,
                // `use crate::foo::*` from within foo.rs) — noise edges
                if file == pf.rel_path {
                    continue;
                }
                let to = format!("file::{}", file);
                let key = (from.clone(), to.clone());
                if !seen.contains(&key) {
                    seen.insert(key);
                    edges.push(crate::contract::edge(&from, &to, "IMPORTS"));
                }
                // rest (an item) is refined by calls/tests via symbol maps
                let _ = rest;
                continue;
            }
            // external crate or unresolved
            let first = crate::thirdparty::first_segment(&u.target);
            if crate::thirdparty::is_std(first) {
                continue; // skip tier — no node, no edge
            }
            if !crate::thirdparty::is_external_crate(first, mf) {
                continue; // unresolvable (typo / not a dep) — metadata only
            }
            // external: [crate]/name (+ member when the use names one) —
            // registration happens inside the registry (gated)
            if tp.crate_node(first).is_some() {
                push_import(
                    &mut edges,
                    &mut seen,
                    &from,
                    &crate::lookup::third_party_id(first),
                );
            }
            let rest: Vec<&str> = u.target.split("::").skip(1).collect();
            if !rest.is_empty() && !u.glob {
                let member = rest.join("::");
                let _ = tp.member_node(first, &member);
            }
        }
    }
    ImportsOutput { edges }
}

fn push_import(
    edges: &mut Vec<CodeEdge>,
    seen: &mut std::collections::HashSet<(String, String)>,
    from: &str,
    to: &str,
) {
    let key = (from.to_string(), to.to_string());
    if !seen.contains(&key) {
        seen.insert(key);
        edges.push(crate::contract::edge(from, to, "IMPORTS"));
    }
}
